pmlc/rt/level_zero/level_zero_invocation.cc

```diff
@@ -0,0 +1,201 @@
+}
+
+LevelZeroKernel::LevelZeroKernel(ze_module_handle_t module, std::string name)
+    : module(module), name(name) {
+  kernel = lzu::create_function(module, 0, name);
+}
+
+LevelZeroKernel::~LevelZeroKernel() {
+  lzu::destroy_function(kernel);
+  lzu::destroy_module(module);
+}
+
+void LevelZeroKernel::addDependency(LevelZeroEvent *event) {
+  dependencies.push_back(event->getEvent());
+}
+
+void LevelZeroKernel::setArg(unsigned idx, LevelZeroMemory *memory) {
+  void *buf = reinterpret_cast<int64_t *>(memory->getBuffer());
+  lzu::set_argument_value(kernel, idx, sizeof(buf), &buf);
+}
+
+void LevelZeroKernel::enqueue(ze_command_list_handle_t list,
+                              ze_group_count_t gws, ze_group_count_t lws,
+                              ze_event_handle_t &resultE) {
+  // Need to find a better way to find right group count.
+  lzu::append_launch_function(list, kernel, &gws, resultE, dependencies.size(),
+                              dependencies.data());
+  // As changes in wrapper, do not need to divide now.
+  // ze_group_count_t groupCount = {
+  //    gws.groupCountX / lws.groupCountX,
+  //    gws.groupCountY / lws.groupCountY,
+  //    gws.groupCountZ / lws.groupCountZ,
+  //};
+  // lzu::append_launch_function(list, kernel, &groupCount, resultE,
+  //                            dependencies.size(), dependencies.data());
+  dependencies.clear();
+}
+
+LevelZeroEvent::LevelZeroEvent(ze_event_handle_t event,
+                               LevelZeroActionKind kind, std::string name)
+    : event(event), kind(kind), name(std::move(name)) {}
+
+void LevelZeroEvent::wait(const std::vector<LevelZeroEvent *> &events) {
+  std::vector<ze_event_handle_t> zeEvents;
+  std::transform(events.begin(), events.end(), std::back_inserter(zeEvents),
+                 [](const LevelZeroEvent *event) { return event->getEvent(); });
+  for (auto e : zeEvents) {
+    zeEventHostSynchronize(e, UINT64_MAX);
+  }
+}
+ze_command_queue_group_properties_t p;
+LevelZeroInvocation::LevelZeroInvocation(LevelZeroDevice *device)
+    : device{device->shared_from_this()}, queueUser(device->getQueue(p)) {
+  // Need a way to ensure event is not used and then can release it
+  eventPool.InitEventPool(device->getLevelZeroContext(), 600);
+
+  // May create queue based on properties once multiple device used
+  // ze_command_queue_group_properties_t p;
+  // p.flags = ZE_COMMAND_QUEUE_GROUP_PROPERTY_FLAG_COMPUTE;
+  // queueUser = device->getQueue(p);
+}
+
+LevelZeroInvocation::~LevelZeroInvocation() {
+  // Need to explicitly wait for all operations to avoid unfinished events
+  // when gathering profiling information.
+  finish();
+
+  for (size_t i = 0; i < memories.size(); i++) {
+    delete memories[i];
+  }
+  for (std::unique_ptr<LevelZeroEvent> &event : events) {
+    eventPool.destroy_event(event->getEvent());
+  }
+  // for (size_t i = 0; i < kernels.size(); i++) {
+  //  delete kernels[i];
```

Comment:
kernels are not used now?

Reply:
comp dialect now create xxx_destroy_kernel to release kernel, so remove these.  But not sure how it will change